Tribler/Test/Community/Tunnel/FullSession/test_tunnel_base.py

```diff
@@ -0,0 +1,191 @@
+
+        # Connect the candidates with each other in all available tunnel communities
+        candidates = []
+        for session in self.sessions:
+            self._logger.debug("Appending candidate from this session to the list")
+            candidates.append(Candidate(session.get_dispersy_instance().lan_address, tunnel=False))
+
+        communities_to_inject = self.tunnel_communities
+        if self.tunnel_community_seeder is not None:
+            communities_to_inject.append(self.tunnel_community_seeder)
+
+        for community in communities_to_inject:
+            for candidate in candidates:
+                self._logger.debug("Add appended candidate as discovered candidate to this community")
+                # We are letting dispersy deal with adding the community's candidate to itself.
+                community.add_discovered_candidate(candidate)
+
+    @blocking_call_on_reactor_thread
+    def load_tunnel_community_in_session(self, session, exitnode=False):
+        """
+        Load the tunnel community in a given session. We are using our own tunnel community here instead of the one
+        used in Tribler.
+        """
+        dispersy = session.get_dispersy_instance()
+        keypair = dispersy.crypto.generate_key(u"curve25519")
+        dispersy_member = dispersy.get_member(private_key=dispersy.crypto.key_to_bin(keypair))
+        settings = TunnelSettings(tribler_session=session)
+        if not self.crypto_enabled:
+            settings.crypto = NoCrypto()
+        settings.become_exitnode = exitnode
+
+        return dispersy.define_auto_load(HiddenTunnelCommunityTests, dispersy_member, (session, settings), load=True)[0]
+
+    @inlineCallbacks
+    def create_proxy(self, index, exitnode=False):
+        """
+        Create a single proxy and load the tunnel community in the session of that proxy.
+        """
+        from Tribler.Core.Session import Session
+
+        self.setUpPreSession()
+        config = self.config.copy()
+        config.set_libtorrent(True)
+        config.set_dispersy(True)
+        config.set_state_dir(self.getStateDir(index))
+
+        session = Session(config, ignore_singleton=True, autoload_discovery=False)
+        session.prestart()
+        yield session.start()
+        self.sessions.append(session)
+
+        returnValue(self.load_tunnel_community_in_session(session, exitnode=exitnode))
+
+    def setup_tunnel_seeder(self, hops):
+        """
+        Setup the seeder.
+        """
+        from Tribler.Core.Session import Session
+
+        self.seed_config = self.config.copy()
+        self.seed_config.set_state_dir(self.getStateDir(2))
+        self.seed_config.set_megacache(True)
+        if self.session2 is None:
+            self.session2 = Session(self.seed_config, ignore_singleton=True, autoload_discovery=False)
+            self.session2.prestart()
+            self.session2.start()
+
+        while not self.session2.lm.initComplete:
+            time.sleep(1)
```

Comment:
@lfdversluis the sleep is not necessary since it's already running on the reactor thread 👍